### DISCUSSION TITLE
What is the best way to mock xero-node in unit tests?

### DISCUSSION THREAD

**Discussion by createthis** [XeroAPI/xero-node#558] [Q&A]
When working with other vendor APIs, there are usually mock packages available to make writing unit tests considerably easier.

### Examples:
- axios
   - axios-mock-adapter
- aws-sdk
   - aws-sdk-mock

I don't see an equivalent for `xero-node`.

What is the best way to mock `xero-node` in unit tests, currently?

**Comment by dimitardanailov** [comment]
@createthis 

Thank you very much for the topic. I'd like to share as an example: https://www.npmjs.com/package/@elastic/elasticsearch-mock

> When testing your application you don't always need to have an Elasticsearch instance up and running, but you might still need to use the client for fetching some data. If you are facing this situation, this library is what you need.

I'm not familiar how is implemented elasticsearch-mock as a package but my expectations are if xero-node has an equivalent to be something like elasticsearch-mock

**Comment by CedricConday** [comment]
Short version: don't mock the `xero-node` classes — mock the HTTP layer underneath. The SDK is a thin wrapper over HTTP, so intercepting there gives realistic tests without coupling to SDK internals. [nock](https://github.com/nock/nock) is the usual Node choice:

```js
import nock from 'nock';

nock('https://api.xero.com')
  .get('/api.xro/2.0/Invoices')
  .reply(200, { Invoices: [{ InvoiceID: '...', Total: 100 }] });

// call xeroClient.accountingApi.getInvoices(tenantId) and assert
```

Tests stay offline and deterministic, and you can assert on the request (URL/headers/body), not just the response. I do the same in my own MCP servers built on fetch-based APIs — contract-style tests over every request with a fetch mock — far less brittle than stubbing client objects, which break whenever the SDK shape changes.

If you do want client-level stubbing, dependency-injecting `XeroClient` and swapping its api objects works too — but HTTP-layer mocking is where I'd start.
